[GStreamer] Harness: Refactor output stream handling in preparation for "sometimes" pad support

philn opened the pull request:
#### 1a42a39e8f4097a91706ddd983c3dd16d8ecdc85
<pre>
[GStreamer] Harness: Refactor output stream handling in preparation for &quot;sometimes&quot; pad support
<a href="https://bugs.webkit.org/show_bug.cgi?id=250757">https://bugs.webkit.org/show_bug.cgi?id=250757</a>

Reviewed by Xabier Rodriguez-Calvar.

The harnessed output stream (corresponding to the element source pad) is now managed through a
dedicated Stream class embedded in the GStreamerElementHarness scope. This new architecture will be
useful for the upcoming support for &quot;sometimes&quot; src pads where the harness will be able to work on
auto-plugger elements such as parsebin and decodebin(3).

No change in functionality, covered by existing API and WebCodecs layout tests.

* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoDecoder::GStreamerInternalVideoDecoder):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
(WebCore::GStreamerElementHarness::GStreamerElementHarness):
(WebCore::GStreamerElementHarness::~GStreamerElementHarness):
(WebCore::GStreamerElementHarness::Stream::Stream):
(WebCore::GStreamerElementHarness::Stream::~Stream):
(WebCore::GStreamerElementHarness::Stream::pullBuffer):
(WebCore::GStreamerElementHarness::Stream::pullEvent):
(WebCore::GStreamerElementHarness::Stream::outputCaps):
(WebCore::GStreamerElementHarness::Stream::chainBuffer):
(WebCore::GStreamerElementHarness::Stream::sinkQuery):
(WebCore::GStreamerElementHarness::Stream::sinkEvent):
(WebCore::GStreamerElementHarness::processOutputBuffers):
(WebCore::GStreamerElementHarness::flush):
(WebCore::GStreamerElementHarness::pullBuffer): Deleted.
(WebCore::GStreamerElementHarness::pullEvent): Deleted.
(WebCore::GStreamerElementHarness::outputCaps): Deleted.
(WebCore::GStreamerElementHarness::chainBuffer): Deleted.
(WebCore::GStreamerElementHarness::sinkQuery): Deleted.
(WebCore::GStreamerElementHarness::sinkEvent): Deleted.
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.h:
(WebCore::GStreamerElementHarness::Stream::create):
(WebCore::GStreamerElementHarness::Stream::pad const):
(WebCore::GStreamerElementHarness::outputStreams const):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GstElementHarness.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/259077@main">https://commits.webkit.org/259077@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a11610ee839d1c7fc502244eba0fe707884e3ad3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12998 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36821 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113108 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173408 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107830 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3889 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96128 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112212 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109652 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93871 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38521 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92640 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25482 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6353 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26864 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6517 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3403 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12506 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46393 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6245 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8287 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
```diff
@@ -208,43 +188,75 @@ bool GStreamerElementHarness::pushEvent(GstEvent* event)
+GStreamerElementHarness::Stream::~Stream()
+{
+    gst_pad_set_active(m_targetPad.get(), FALSE);
+    {
```
This scope is not needed. There is nothing below that scope so it is going to unlock anyway.

philn committed:
[GStreamer] Harness: Refactor output stream handling in preparation for "sometimes" pad support
https://bugs.webkit.org/show_bug.cgi?id=250757

Reviewed by Xabier Rodriguez-Calvar.

The harnessed output stream (corresponding to the element source pad) is now managed through a
dedicated Stream class embedded in the GStreamerElementHarness scope. This new architecture will be
useful for the upcoming support for "sometimes" src pads where the harness will be able to work on
auto-plugger elements such as parsebin and decodebin(3).

No change in functionality, covered by existing API and WebCodecs layout tests.

* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoDecoder::GStreamerInternalVideoDecoder):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
(WebCore::GStreamerElementHarness::GStreamerElementHarness):
(WebCore::GStreamerElementHarness::~GStreamerElementHarness):
(WebCore::GStreamerElementHarness::Stream::Stream):
(WebCore::GStreamerElementHarness::Stream::~Stream):
(WebCore::GStreamerElementHarness::Stream::pullBuffer):
(WebCore::GStreamerElementHarness::Stream::pullEvent):
(WebCore::GStreamerElementHarness::Stream::outputCaps):
(WebCore::GStreamerElementHarness::Stream::chainBuffer):
(WebCore::GStreamerElementHarness::Stream::sinkQuery):
(WebCore::GStreamerElementHarness::Stream::sinkEvent):
(WebCore::GStreamerElementHarness::processOutputBuffers):
(WebCore::GStreamerElementHarness::flush):
(WebCore::GStreamerElementHarness::pullBuffer): Deleted.
(WebCore::GStreamerElementHarness::pullEvent): Deleted.
(WebCore::GStreamerElementHarness::outputCaps): Deleted.
(WebCore::GStreamerElementHarness::chainBuffer): Deleted.
(WebCore::GStreamerElementHarness::sinkQuery): Deleted.
(WebCore::GStreamerElementHarness::sinkEvent): Deleted.
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.h:
(WebCore::GStreamerElementHarness::Stream::create):
(WebCore::GStreamerElementHarness::Stream::pad const):
(WebCore::GStreamerElementHarness::outputStreams const):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GstElementHarness.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/259077@main
